Add depUrls mechanism and update processZlib to optionally remote load pako

c42f373607bf2b opened the pull request:
node's zlib is not available in the WebIDE, so `process: zlib` doesn't work there.
<strike>
Switch out zlib for [pako.js](https://github.com/nodeca/pako) which works in both browser
and node. Author claims it's nearly as fast as c zlib, so there are no perf concerns.

Will open Separate PR in WebIDE repo, to make sure pako is deployed to the
proper asset path on the server.

This **will break** a WebIDE deploy if it isn't upgraded with the corresponding  https://github.com/kaitai-io/kaitai_struct_webide/pull/46
</strike>

**Update 1**: allow user to configure a new `depUrls` object with a url from which to load [pako.js](https://github.com/nodeca/pako), and use it to do zlib decompression in the browser.

GreyCat commented on package.json:
```diff
@@ -7,6 +7,9 @@
     "type": "git",
     "url": "git+https://github.com/kaitai-io/kaitai_struct_javascript_runtime.git"
   },
+    "dependencies": {
```
Looks like a formatting problem here.

c42f373607bf2b replied:
removed dep

GreyCat commented:
I'm not a big fan of adding more hard dependencies to JS runtime. The problems I see:

1. zlib is not really used by 95-98% of the users => extra dependency is just waste of resources / time / complicates deployment.
2. Whenever it is used in node.js environment, native `zlib` should be used (it is always available, it's faster, no extra effort for dependencies).
3. In other environments (i.e. browser) pako indeed proves to be useful, but that should be extra option, not a sole solution.

Thus, I'd propose to either do some sort of "soft" dependency — i.e. without clause in package.json, but with some sort of autodetection in `processZlib`, like the stuff we're doing with `iconv-lite`. Of course, it should be documented somewhere around [JavaScript notes](http://doc.kaitai.io/lang_javascript.html), so people would know which options are available. If autodetection is not possible, may be we should add some machinery to configure it before usage, i.e. something like

```javascript
KaitaiStream.zlibEngine = "pako";
KaitaiStream.zlibEngine = "zlib";
```

GreyCat commented on KaitaiStream.js:
```diff
@@ -562,14 +562,18 @@ KaitaiStream.processRotateLeft = function(data, amount, groupSize) {
+  var r;
+  if (typeof require === 'undefined')  {
+    // no require() -> Assume we're running as a webworker in the browser
+    importScripts("../../lib/_npm/pako/pako_inflate.min.js");
```
This seems to refer some paths inside web IDE, and runtime definitely must not include any hard codes tailored specifically for web IDE. Lots of people use JS runtime in a browser outside of web IDE, and we shouldn't force their projects to use some specific paths or whatever.

c42f373607bf2b replied:
made configurable

c42f373607bf2b commented:
Ok, here's the new iteration addressing your concerns:

1. hard dep removed from package.json
2. KaitaiStream gained a "depUrls" property - assuming future additions will require other dependencies (crypto libs for example), this is intended to be a uniform configuration mechanism
3. if running in node (`require()` is present), the previous code and node's 'zlib' is used
4. if `require()` is not present, and depUrls.zlib has been configured by the user (externally),
importScripts() is called.
5. The WebIDE PR has been updated. The hard dependency on pako is still there, and now
`depUrls.zlib` is configured as well.

Let me if this solution works for you and I'll open a follow up PR with docs.

GreyCat commented:
@c42f373607bf2b I'm sorry for not-so-hasty review, I keep having some health problems, so I'm kind of slow :(

> anticipating the need for additional soft dependencies by future enhancements (crypto libs, for example), this new object is intended to serve as the designated store for dependency information

This is actually an interesting question that we've kind of tried to solve in https://github.com/kaitai-io/kaitai_struct/issues/51. Nowadays, it's possible to do stuff like:

```yaml
- id: foo
  size: 0x1000
  process: kaitai.crypto.aes192(some_iv, some_key)
```

which will translate into a call like that in JS:

```javascript
(function (root, factory) {
  if (typeof define === 'function' && define.amd) {
    define(['kaitai-struct/KaitaiStream', 'kaitai-crypto/Aes192'], factory);
  } else if (typeof module === 'object' && module.exports) {
    module.exports = factory(require('kaitai-struct/KaitaiStream'), require('kaitai-crypto/Aes192'));
  } else {
    root.Aaaa = factory(root.KaitaiStream, root.Aes192);
  }
}(this, function (KaitaiStream, Aes192) {
// ...
    this._raw_foo = this._io.readBytes(4096);
    var _process = new Aes192(this.someIv, this.someKey);
    this.foo = _process.decode(this._raw_foo);
```

So, basically, it should be meant to be extensible alternative for adding lots of other compression/crypto methods, and, to be honest, given the amount of dependencies zlib brings in some cases, I've thought about deprecating zlib functionality in stdlib and doing it as external module as well.

But, anyway, your solution closes the existing problem, and I believe it's ok to start merging it. @koczkatamas, do you have any concerns?

c42f373607bf2b commented:
> I'm sorry for not-so-hasty review, I keep having some health problems, so I'm kind of slow :(

Be well... :thermometer: : 

> This is actually an interesting question

and a really long thread...

A few extra points for discussion:

- Should parameterization work via function naming instead of arguments (`aes192(foo)` instead of `aes(192,foo)`? it means the number of functions is exponential in the number of params: (aes X [128,192,156] X [`cbc`,`otr`,`eax`,`gcm`])... Seems like it may produce lots of boilerplate.

- The Process functions are static, i.e. there's no `this` to hang the config data on. That means that we either put the config data on a global (the factory function `KaitaiStream` in this PR), or pass it in as an argument at each invocation. Possibly to each call to init/process/finalize. For one thing it means that can only configure a module once. If you need another configuration, you need a different module.
That makes it a little tricky to decide what should be a call parameter and what should be configuration when writing new modules. For something like `depUrls` that's probably ok, but is it enough for for future cases? 

- Just having `depUrls` seems too specific. `depUrls` should be replaced with something more generic like `config.zlib.scriptUrl`. Specifically, the configuration should be grouped by module.

I don't want to go too far with over-engineering, these are just some points to consider before choosing a way forward.

c42f373607bf2b commented:
@GreyCat , ping

koczkatamas commented:
Well, with this solution zlib cannot be used in browser's DOM context, only in web workers, but as `processZlib` is a sync function, currently there is no solution for this problem (AFAIK sync file download is not possible anymore, but even it is possible, we should not use that).

So this is the best we can. In the long run preloading pako before parsing with an UMD-style dependency loading would be better, but till then this PR will solve some major problems with zlib, so LGTM.

GreyCat commented:
I've merged this, so hopefully it will solve web IDE problems ATM. I'll try to fix tests issues soon (if there would be any).

> Should parameterization work via function naming instead of arguments (`aes192(foo)` instead of `aes(192,foo)`)? it means the number of functions is exponential in the number of params: (aes X [128,192,156] X [`cbc`,`otr`,`eax`,`gcm`])... Seems like it may produce lots of boilerplate.

This is really up to crypto library API designer's to decide. If I were doing that, I'd analyze existing implementation for major platforms/languages and base that on how it's done there. If there are distinct constants or something for all these permutations — then probably that's how we should do it.

> The Process functions are static

It's kind of hard to guess now, I propose to start developing stuff and fix our model later, if it would be needed (i.e. application-level config, call-level config, ksy-level config, etc).
